### PR TITLE
feat(Transfer): support multiple selection by holding down the shift key

### PR DESCRIPTION
**Pull request by zbw-zbw**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
https://github.com/ant-design/ant-design/issues/44669

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     Transfer support multiple selection by holding down the shift key      |
| 🇨🇳 Chinese |     Transfer 穿梭框组件支持按住 shift 键多选      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at bc917c6</samp>

Added multiple selection feature and event parameter to the transfer component and its demos. Fixed a bug in the table transfer demo and improved the code style and consistency of the import statements and prop functions. Updated the documentation to reflect the new API.

### <samp>🤖 Generated by Copilot at bc917c6</samp>

*  Add `event` parameter of type `React.MouseEvent` to `onItemSelect` prop function in transfer component and update documentation ([link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL76-R77), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-edca56351272e05c831757bdcbfe5641e3d7615af8bfaa391b4ad9ef8b9eb49cL67-R69), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-a3fcfc2351816508d8b94df43d29a886ee240f42726572e193634aa2ce09ce75L67-R69), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-5d2a9fbb1586f5058bfd77d20e281f359ce068ed333ccc4375ee0ef1961342cbL59-R64), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-c165dbeb24339c042fcc975a9b661af8947ea942457cf8935f29817ac86033a9L71-R78), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdL283-R391), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-f25eb82686b163c8fb19e3d453b3c168ce5311046d76e7b3a1192173b676ea97L74-R81), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415L55-R56))
*  Implement multiple selection by holding down the shift key in transfer component using refs to store the previous selected item index and helper functions to handle single and multiple selection scenarios ([link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdR172-R175), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdR200-R208), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdL239-R257), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdR280), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdL283-R391))
*  Add test case for multiple selection by holding down the shift key in transfer component ([link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296R155-R176))
*  Adjust import order of modules in `./components/transfer/ListBody.tsx`, `./components/transfer/ListItem.tsx`, `./components/transfer/__tests__/index.test.tsx`, `./components/transfer/demo/component-token.tsx`, `./components/transfer/demo/table-transfer.tsx`, `./components/transfer/index.tsx`, and `./components/transfer/list.tsx` to follow the convention of placing React as the first import, followed by external modules, then internal modules, and then type imports ([link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL1-R12),  [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296L1-R4), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-edca56351272e05c831757bdcbfe5641e3d7615af8bfaa391b4ad9ef8b9eb49cL1-R5), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-a3fcfc2351816508d8b94df43d29a886ee240f42726572e193634aa2ce09ce75L2-R5), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdL2-R2), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415L1-R5))
*  Move imports of `ListBodyRef`, `TransferListBodyProps`, `DefaultListBody`, and `OmitProps` from `./components/transfer/ListBody` to the bottom of the file in `./components/transfer/list.tsx` to reduce the scope of these imports and avoid potential circular dependencies ([link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415L10-L11), [link](https://github.com/ant-design/ant-design/pull/44798/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415R20-R21))

### CODE REVIEW

File: components/transfer/__tests__/index.test.tsx [hunk 1]
```diff
@@ -151,6 +152,65 @@ describe('Transfer', () => {
...
+    );
+
+    fireEvent.click(getByText('b'));
+    fireEvent.click(getByText('d'));
+    fireEvent.click(getByText('f'), {
+      shiftKey: true,
+    });
+    expect(handleSelectChange).toHaveBeenLastCalledWith([], ['b', 'd', 'e', 'f']);
+
+    fireEvent.click(getByText('f'));
```
Comment (MadCcc) [inline components/transfer/__tests__/index.test.tsx:207]: 这一步反选是不是结果上 f 不应该被选中？

Comment (zbw-zbw) [reply components/transfer/__tests__/index.test.tsx:207]: 对，就反选f这一步f是没选中的，我这里没输出结果，因为反选完f紧接着选中了g，所以最终结果f是选上的

Comment (zbw-zbw) [reply components/transfer/__tests__/index.test.tsx:207]: 反选完的结果应该是这样，看看是不是要补一下这里的结果？
image

Comment (zbw-zbw) [reply components/transfer/__tests__/index.test.tsx:213]: done

**Comment by MadCcc** [comment]
拿 finder 试了一下 shift 的逻辑，感觉还挺复杂。。

**Comment by MadCcc** [comment]
这个能不能和 Table 那边一起抽一个 hook 出来？

**Comment by zbw-zbw** [comment]
> 拿 finder 试了一下 shift 的逻辑，感觉还挺复杂。。

嗯嗯 我在做之前也看过 finder 的效果， 在某些场景下我们 transfer 的效果跟它的不太一样（finder 是单选，按住  command 才能单击选中多个，我们是可以持续单击进行多选的场景）所以大部分场景是跟 finder 效果差不多

**Comment by zbw-zbw** [comment]
> 拿 finder 试了一下 shift 的逻辑，感觉还挺复杂。。

具体效果也可以本地运行看下哈～

**Comment by zbw-zbw** [comment]
> 这个能不能和 Table 那边一起抽一个 hook 出来？

这个要看下 Table 那边的实现逻辑 ，因为 transfer 是左右两个列表分开的 分别支持多选，我有空看看 👀～

**Comment by zbw-zbw** [comment]
> 这个能不能和 Table 那边一起抽一个 hook 出来？
> 
> 

刚看了下 table 的效果和逻辑都跟我的实现都不太一样耶哈哈哈 😳 table 那边取消选中之后就不能 shift 多选了 要先单选一个才能多选 我这边可以 耶✌🏻️

**Comment by MadCcc** [comment]
这种操作应该是要统一一下的，尽量复用

**Comment by zbw-zbw** [comment]
> 这种操作应该是要统一一下的，尽量复用

嗯嗯好 是要以 table 的为准嘛？还是以最新的实现效果为准

**Comment by zbw-zbw** [comment]
> 这种操作应该是要统一一下的，尽量复用

done

File: components/table/hooks/useSelection.tsx [hunk 2]
```diff
@@ -109,6 +110,10 @@ const useSelection = <RecordType extends AnyObject = AnyObject>(
 
   const warning = devUseWarning('Table');
 
+  // ========================= MultipleSelect =========================
+  const prevSelectedIndexRef = React.useRef(-1);
```
Comment (zombieJ) [inline components/table/hooks/useSelection.tsx:114]: 看起来不应该用 ref 来管，useState 会更好。
另外如果你是提供了一个 hook 做全选，那么所有逻辑都应该在 hook 里，而不是单独外面放一个 hook

Comment (zbw-zbw) [reply components/table/hooks/useSelection.tsx:114]: > 看起来不应该用 ref 来管，useState 会更好。 另外如果你是提供了一个 hook 做全选，那么所有逻辑都应该在 hook 里，而不是单独外面放一个 hook

+ 确实应该用 useState 管理
+ OK 这里写法确实有问题，我改造下代码结构

Comment (zbw-zbw) [reply components/table/hooks/useSelection.tsx:114]: > 看起来不应该用 ref 来管，useState 会更好。 另外如果你是提供了一个 hook 做全选，那么所有逻辑都应该在 hook 里，而不是单独外面放一个 hook

仔细想了下，好像用 ref 管理好一点或者说方便一点，prevSelectedIndexRef 为了及时记录/更新上次选中的下标，不涉及视图状态的变化

**Comment by yoyo837** [comment]
push -f

**Pull request closed without merging**